### PR TITLE
[NHUB-570] Upgrade Monitoring resource and views to async

### PR DESCRIPTION
**Pull request by MarkLark86**
### Purpose
Upgrade the Monitoring and associated functionalities to async

### What has changed
* New MonitoringProfile resource & service
* New MonitoringSearch service (inherits from WireSearch service)
* Registers new module
* Update email_alert command to a class and use async resources
* Update Monitoring formatters to async, and add types
* Update Monitoring views to use async resources
* Fix types with Newshub service and EmailKwargs

Resolves: [NHUB-570](https://sofab.atlassian.net/browse/NHUB-570)

[NHUB-570]: https://sofab.atlassian.net/browse/NHUB-570?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: newsroom/monitoring/email_alerts.py [hunk 1]
```diff
@@ -9,38 +9,46 @@
...
+class MonitoringEmailAlerts:
+    def __init__(self):
         self.log_msg = f"Monitoring Scheduled Alerts: {utcnow()}"
 
-    async def run(self, immediate=False):
+    async def run(self, immediate: bool = False) -> None:
         try:
-            # TODO-ASYNC: update this below when monitoring is migrated to async
-            get_resource_service("monitoring")
+            get_current_async_app().resources.get_resource_service("monitoring")
```
Comment (eos87) [inline newsroom/monitoring/email_alerts.py:51]: circular reference or modules loading race condition?

File: newsroom/monitoring/filters.py [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+
+def prefill_search_section(request: NewshubSearchRequest[MonitoringSearchRequestArgs]) -> None:
+    """Set the request section to WIRE, as that's where the content is"""
+    request.section = SectionEnum.WIRE
+
+
+async def prefill_monitoring_user(request: NewshubSearchRequest[MonitoringSearchRequestArgs]) -> None:
+    """Prefill the user from the current request details"""
+
+    if request.args.skip_user_validation:
```
Comment (eos87) [inline newsroom/monitoring/filters.py:27]: out of curiosity, what's the reason to skip the user validation here?

File: newsroom/monitoring/views.py [hunk 3]
```diff
@@ -83,103 +79,120 @@ def process_form_request(updates, request_updates, form):
...
+
 
+class MonitoringIdUrlArg(BaseModel):
+    id: ObjectId
 
-@blueprint.route("/monitoring/<id>/users", methods=["POST"])
-@account_manager_only
-async def update_users(id):
+
+@monitoring_endpoints.endpoint("/monitoring/<id>/users", methods=["POST"], auth=[auth_rules.account_manager_only])
```
Comment (eos87) [inline newsroom/monitoring/views.py]:
```suggestion
@monitoring_endpoints.endpoint("/monitoring/<TYPE:id>/users", methods=["POST"], auth=[auth_rules.account_manager_only])
```
Should we have the type here?

File: newsroom/monitoring/views.py [hunk 4]
```diff
@@ -83,103 +79,120 @@ def process_form_request(updates, request_updates, form):
...
+        [monitoring.company async for monitoring in cursor if monitoring.company]
+    )
 
+    return Response(companies)
 
-@blueprint.route("/monitoring/<id>/schedule", methods=["POST"])
-@account_manager_only
-async def update_schedule(id):
+
+@monitoring_endpoints.endpoint("/monitoring/<id>/schedule", methods=["POST"], auth=[auth_rules.account_manager_only])
```
Comment (eos87) [inline newsroom/monitoring/views.py]: same here?

```suggestion
@monitoring_endpoints.endpoint("/monitoring/<TYPE:id>/schedule", methods=["POST"], auth=[auth_rules.account_manager_only])
```

File: newsroom/monitoring/views.py [hunk 5]
```diff
@@ -188,86 +201,111 @@ async def edit(_id):
...
+            await MonitoringProfileService().update(args.item_id, updates)
+            return Response({"success": True})
+        return Response(form.errors, 400)
+    return Response(profile)
 
 
-@blueprint.route("/monitoring/<_id>", methods=["DELETE"])
-@admin_only
-async def delete(_id):
+@monitoring_endpoints.endpoint("/monitoring/<_id>", methods=["DELETE"], auth=[auth_rules.admin_only])
```
Comment (eos87) [inline newsroom/monitoring/views.py]:
```suggestion
@monitoring_endpoints.endpoint("/monitoring/<string:id>", methods=["DELETE"], auth=[auth_rules.admin_only])
```

File: newsroom/monitoring/views.py [hunk 6]
```diff
@@ -188,86 +201,111 @@ async def edit(_id):
...
-    monitoring_profile["alert_type"] = layout_format
-    items = get_items_for_monitoring_report([_id for _id in _ids.split(",")], monitoring_profile)
+
+class ExportMonitoringUrlParams(BaseModel):
+    monitoring_profile: ObjectId
+    format: str | None = None
+    secondary_format: str | None = None
+
+
+@monitoring_endpoints.endpoint("/monitoring/export/<_ids>")
```
Comment (eos87) [inline newsroom/monitoring/views.py]:
```suggestion
@monitoring_endpoints.endpoint("/monitoring/export/<string:_ids>")
```

File: newsroom/monitoring/views.py [hunk 7]
```diff
@@ -188,86 +201,111 @@ async def edit(_id):
...
+async def export(args: ExportMonitoringUrlArgs, params: ExportMonitoringUrlParams, request: Request) -> Response:
+    user = get_user_from_request(request)
+    if not params.format:
+        return Response({"message": gettext("No format specified")}, 400)
+
+    formatter = get_current_app().as_any().download_formatters[params.format]["formatter"]
```
Comment (eos87) [inline newsroom/monitoring/views.py]: if you use `newsroom.core.get_current_wsgi_app` you won't need the `as_any`

```suggestion
    formatter = get_current_wsgi_app().download_formatters[params.format]["formatter"]
```

**Review by eos87 — APPROVED** [review]
Looks, just left a few minor comments.

File: newsroom/monitoring/email_alerts.py (hunk 1 above)
Comment (MarkLark86) [reply newsroom/monitoring/email_alerts.py:51]: Just an easy way to determine if the Monitoring module is loaded (as it's an optional feature to Newshub)

File: newsroom/monitoring/filters.py (hunk 2 above)
Comment (MarkLark86) [reply newsroom/monitoring/filters.py:27]: Monitoring emails are sent based on Company permissions, so we skip user validation

**Pull request merged**